greynoise_connector.py

```diff
@@ -161,438 +148,353 @@ def _is_valid_ip(self, input_ip_address):
-
-                if full_response["count"] <= item_size:
-                    action_result.add_data(full_response)
-
-                # if there are more results than what is requested delete them from the query results
-                if full_response["count"] > item_size:
-                    temp = []
-                    for i in range(item_size):
-                        temp.append(full_response["data"][i])
-                    del full_response["data"]
-                    del full_response["count"]
-                    full_response["data"] = temp
-                    full_response["count"] = len(temp)
-                    action_result.add_data(full_response)
-
-                try:
-                    for entry in full_response["data"]:
-                        entry["visualization"] = VISUALIZATION_URL.format(ip=entry["ip"])
-                except KeyError:
-                    error_message = "Error occurred while processing API response"
-                    return action_result.set_status(phantom.APP_ERROR, error_message)
+                    return action_result.set_status(phantom.APP_ERROR, message)
+                if is_poll:
+                    return results
+                else:
+                    if results.get("data", []):
+                        for ip_info in results.get("data", []):
+                            trust_level = ip_info.get("business_service_intelligence", {}).get("trust_level", "")
+                            ip_info["trust_level"] = TRUST_LEVELS.get(str(trust_level), trust_level)
+                            ip_info["visualization"] = VISUALIZATION_URL.format(ip=ip_info["ip"])
+                    action_result.add_data(results)
+            except KeyError:
+                error_message = API_PARSE_ERROR_MESSAGE
+                return action_result.set_status(phantom.APP_ERROR, error_message)
         except Exception as e:
             error_message = self._get_error_message_from_exception(e)
             return action_result.set_status(phantom.APP_ERROR, urllib.parse.unquote(error_message))
 
-        return action_result.set_status(phantom.APP_SUCCESS, "GNQL Query action successfully completed")
+        return action_result.set_status(phantom.APP_SUCCESS, GNQL_QUERY_SUCCESS_MESSAGE)
 
-    def _lookup_similar_ips(self, param):
-        self.save_progress(GREYNOISE_ACTION_HANDLER_MESSAGE.format(identifier=self.get_action_identifier()))
-        action_result = self.add_action_result(ActionResult(dict(param)))
+    def _paginator(self, query, size=None, exclude_raw=False, quick=False):
+        """Fetch paginated results for a GreyNoise query."""
+        query_response = {"data": []}
+        message = "Successfully fetched query data"
 
-        license_type, message = self._check_license_type()
-        if license_type == "community":
-            return action_result.set_status(phantom.APP_ERROR, message)
+        # Calculate optimal page size
+        page_size = 1000
```

Comment:
This should probably be 10000.  Any request that includes a limit under 10k can be done in a single API call.  It is only if more than 10k results are requested that you would need to paginate the results.